snap: add support for kernel-gpu-2404

Saviq opened the pull request.

Saviq committed:
new: add component-monitor

Add a component monitor that will manage content in the target path
based on the contents of a sentinel file under the source path.

Saviq committed:
snap: add kernel-gpu-2404 content support

Using hooks and the component monitor, ensure the content is managed in
${SNAP_DATA}/kernel-gpu-2404

Saviq committed:
snap: use the kernel component wrapper

Check for sentinel existence in the second content location and run
through the wrapper within.

TheSignPainter98 commented on component-monitor/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+
+[dependencies]
+clap = { version = "4", default-features = false, features = ["derive", "env", "help", "std", "usage"] }
+exitcode = { version = "1", default-features = false }
```
I think this dependency can be safely dropped, it looks like it's just being used to get `exitcode::OK`, which I think can be equivalently expressed using [`std::process::ExitCode::SUCCESS`](https://doc.rust-lang.org/std/process/struct.ExitCode.html#associatedconstant.SUCCESS)

Saviq replied:
I tried to get this going, but I've found no accepted way to convert `::SUCCESS` to `i32` that `::exit()` wants…

Went with literal `0` in the end.

TheSignPainter98 commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+    thread::spawn(move || {
+        for _sig in signals.forever() {
+            std::process::exit(exitcode::OK);
+        }
```
The use of a loop here feels slightly odd as this as `signals.forever()` is actually being used as a fence and the body of the loop is only expected to be run once. Perhaps something like this could be clearer?
```rust
signals.forever().next(); // Await signal.
process::exit(ExitCode::SUCCESS);
```

TheSignPainter98 commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+    };
+
+    let sentinel_tgt = target.join(sentinel);
+    if sentinel_data == fs::read(&sentinel_tgt).unwrap_or_default() {
```
The use of `.unwrap_or_default()` feels a little unclean here as it's being used to ignore the error case and create a new string. This check also relies on the condition that valid `sentinel_data != ""` in order to work, and although this may always be the case, it's an extra precondition which must be kept in mind. I think instead, it could be clearer to use something like this to explicitly ignore the error and then check the content if successfully read

```rust
fs::read(&sentinel_tgt).is_ok_and(|content| content == sentinel_data)
```

TheSignPainter98 commented on component-monitor/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+edition = "2021"
+
+[dependencies]
+clap = { version = "4", default-features = false, features = ["derive", "env", "help", "std", "usage"] }
```
I'm a little surprised that the `color`, `error-context` and `suggestions` features are absent, is this for compatibility? Personally I'm a sucker for pretty terminal applications :D (have you also considered the `wrap_help` feature?) I see the rest of the dependencies deliberately don't have their default features enabled so I suspect this is all intentional

Saviq replied:
Really it's just for keeping it tight. It's not really a CLI application, just an internal service.

I dropped `default_features` across the board b/c I dislike things creeping up on me, is all ;)

And Innes warned me about `nix` being heavy, so that's probably why ;)

TheSignPainter98 replied:
Ah, this sounds really good, thanks for the clarification!

TheSignPainter98 commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+    #[arg(env = "COMPONENT_SOURCE")]
+    source: PathBuf,
+    /// The sentinel file name
+    #[arg(env = "COMPONENT_SENTINEL", value_parser = clap::builder::NonEmptyStringValueParser::new())]
```
I think there are many ways the user can get a sentinel incorrect and given that it's quite difficult to pass an empty string here, I think it's probably okay not to check this explicitly.

_However,_ if there is a reasonable finite number of inotify events we wish to support here, we could use an enum and implicitly get better validation:

```rust
use clap::ValueEnum;

#[derive(ValueEnum)]
// #[value(rename_all = "SCREAMING_SNAKE_CASE")] // Default is kebab-case, not sure
                                                 // if this is more appropriate?
enum Sentinel {
    InAccess,
    InModify,
    ..
}

Saviq replied:
The sentinel here is actually a file name that we're monitoring for… I'm not reading in inotify events as arguments?

Or did I get my streams crossed here?

TheSignPainter98 replied:
_Oooooh,_ right, when I first read this, I thought `sentinel` here related to the inotify events! I think my confusion came from the use of the `String` type here---it looked like this specificially _isn't_ a file given that `target` has type `PathBuf` below. I think using `PathBuf` for both would clean this up nicely. 

For reference, `String` should be avoided for buffers as Rust's `String` and `&str` types both require valid UTF-8, so to drop this requrement, `PathBuf` is more idiomatic

Saviq replied:
I suppose I thought that `Path*` would not be appropriate for just a file name… but of course a file name is still a path…

TheSignPainter98 replied:
Ah, but the string could still contain a `/` making it more like a path :) I think the intention is to use this with the file system, so signalling that could be beneficial (hence `Path*`).

This actually makes me wonder, if `sentinel` is expected not to contain any path separators, rather than having separate target and sentinel parameters, could we take one and extract the name of the directory? Otherwise, I think `sentinel` feels more like a path to me

Saviq replied:
> could we take one and extract the name of the directory?

I suppose if we define the source folder as the parent of the sentinel, that makes sense. And indeed avoids some rabbit holes.

TheSignPainter98 commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+            Err(e) => {
+                error!(target: "files", "{e}");
+            }
+        }
```
This section is quite deeply nested, I think it would be cleaner to flatten this with by assigning out of the outermost `match`, adding a few `continue`s and then cleaning things up a little with `if-let` where we only care about the data in one branch of the match (this last change is a little more subjective than the others). How about something like this?

```rust
for entry in ... {
    let entry = match entry {
        Ok(e) => e,
        Err(err) => {
            error!(target: "files", "{err}");
            continue;
        },
    };
    
    let entry_path = entry.path();
    if entry_path.is_dir() {
        if let Err(err) = fs::remove_dir_all(&entry_path) {
            error!(..., "failed to remove ...");
            continue;
        }
        debug!(..., "removed ... recursively");
        continue;
    }

    if let Err(err) = fs::remove_file(&path) {
        error!(..., "failed to remove ...");
        continue;
    }
    debug!(..., "removed ...")
}

TheSignPainter98 commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+
+/// Check if the sentinel file is current, else clean the target directory and populate again,
+/// sentinel file being last
+fn populate(source: &PathBuf, sentinel: &String, target: &PathBuf) {
```
The types here are slightly non-idiomatic, specifically, we're passing immutable references to the owned versions of each of these types (the owned versions have enough information to perform mutation). This is a little awkward as if, for example, we wanted to pass a string literal here---it would force that `&'static str` to be copied into a `String` before it could be used! Instead, it's better practice to use to take references to the non-owned versions of each of these, i.e. `&Path` and `&str`. The compiler will automatically coerce these `&Owned`  types into their `&Borrowed` equivalents, so this change will make this signature simpler without any real cost

Saviq committed:
kernel-gpu: make the component precede Mesa

Since the component is likely to be more appropriate for the hardware,
make it go first.

TheSignPainter98 commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+                InotifyEvent {
+                    mask: AddWatchFlags::IN_DELETE_SELF | AddWatchFlags::IN_MOVE_SELF,
+                    ..
+                } => panic!("ERROR: Monitored folder disappeared"),
```
This seems like quite an easy panic to hit! Panicking is a very aggressive action and there is no guarantee that resources are nicely cleaned up (if this program were compiled with `panic = "abort"` set in Cargo.toml e.g. to reduce binary size). I think it would be nicer just to log an error and return here

TheSignPainter98 reviewed (dismissed):
This is looking nice! I think this PR could use a little extra error handling, for which I provide some suggestions :)

Edit: apparently my suggestions on error handling got dropped from my review... thanks, GitHub! Anyway, my suggestion would be to use [`anyhow`](https://docs.rs/anyhow/latest/anyhow/)

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+use log::{debug, error, info, warn, LevelFilter};
+use nix::sys::inotify::{AddWatchFlags, InitFlags, Inotify, InotifyEvent};
+use signal_hook::{consts::SIGINT, consts::SIGTERM, iterator::Signals};
```
just a suggestion

```suggestion
use signal_hook::{consts::{SIGINT, SIGTERM}, iterator::Signals};
```

paul-rodriguez commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+                if path == source {
+                    continue;
+                }
+                if *(relative_path.to_str().unwrap()) == **sentinel {
```
There should be a way to get this to read `relative_path == sentinel.into()` and then compare them as `PathBuf`

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+    info!(target: "files", "copying from {source:?} to {target:?}");
+    for entry in WalkDir::new(&source) {
+        match entry {
+            Ok(e) => {
```
(from @TheSignPainter98) Instead of having the whole logic nested, we can use here something like this:

```rust
for entry in ... {
    let entry = match entry {
        Ok(e) => e,
        Err(err) => {
            error!(target: "files", "{err}");
            continue;
        },
    };

    let entry_path = entry.path();
    if entry_path.is_dir() {
        if let Err(err) = fs::remove_dir_all(&entry_path) {
            error!(..., "failed to remove ...");
            continue;
        }
        debug!(..., "removed ... recursively");
        continue;
    }

    if let Err(err) = fs::remove_file(&path) {
        error!(..., "failed to remove ...");
        continue;
    }
    debug!(..., "removed ...")
}
```

Then you can use the value in the code, avoiding additional nesting

TheSignPainter98 replied:
Oh weird, apparently my comment above _did_ make it into the review?! Amazing, thanks for making sure this is seen anyway Nadzeya!

paul-rodriguez commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+        match entry {
+            Ok(e) => {
+                let path = e.path();
+                let relative_path = path.strip_prefix(source).unwrap();
```
You should try to avoid unwrap and prefer `?` instead

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+
+    cleanup(target, sentinel);
+    info!(target: "files", "copying from {source:?} to {target:?}");
+    for entry in WalkDir::new(&source) {
```
Suggestion from `cargo clippy`: change this to: `source`

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,238 @@
+                    name: Some(name),
+                    mask,
+                    ..
+                } if *name == *args.sentinel => match mask {
```
(from @TheSignPainter98 and @sminez)
Since `InotifyEvent` is not an enum but a struct, using `match` here is unexpected. It's better to extract the values and match them separately:

```rust
let (event_name, mask) = match event { InotifyEvent { name, mask .. } => (name, mask),
    _ => panic!("..."),
};
```

Moreover, since it seems that you need only the fields from this struct, you can simplify it the following way:

```rust
for InotifyEvent { name, mask, .. } in events {
    ...
}
```

Saviq reviewed (commented):
Thanks all for the comments!

Will reply for now to just the ones I have any excuse for :X

Saviq reviewed (commented):
Hmm how did these not get in…

Saviq committed:
snap: move mangler out of `bin/`

Saviq committed:
Review #1: drop exitcode, use paths, next on signals

- drop `exitcode` dependency
- use `PathBuf` for sentinel name
- use `.next()` for signals rather than looping

Co-authored-by: Ed Jones <ed.jones@canonical.com>

Saviq committed:
Review #2: redo error handling

Co-authored-by: Ed Jones <ed.jones@canonical.com>
Co-authored-by: Paul Rodriguez Lobera <paul.rodriguez@canonical.com>

Saviq committed:
Review #3: clippy tweaks, refactor Inotify events

- tweaks to `use`
- clippy suggestions
- refactor interpreting Inotify events

Co-authored-by: Nadzeya Hutsko <nadzeya.hutsko@canonical.com>

Saviq committed:
Review #4: refactor file tree handling

Co-authored-by: Ed Jones <ed.jones@canonical.com>

Saviq committed:
Review #5: extract sentinel name from a full path

Saviq committed:
cleanup: avoid unnecessary placeholder ids

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,242 @@
+}
+
+/// Remove the sentinel file from target, then clear the target directory
+fn cleanup(target: &PathBuf, sentinel_name: &str) -> Result<()> {
```
Instead of using `PathBuf` for function arguments, you can prefer `Path`
`Path` is more lightweight, and you can convert a `PathBuf` to a `Path` when needed with `.as_path()`.

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,242 @@
+    }
+
+    let sentinel_data = match fs::read(&sentinel_src) {
+        Ok(d) if d.trim_ascii() == b"" => {
```
I guess you can try `is_empty` instead  
```suggestion
        Ok(d) if d.is_empty() => {
```

Saviq replied:
I don't want whitespace to be considered, hence the `trim_ascii()`.

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,242 @@
+    thread::spawn(move || {
+        signals.forever().next();
+        std::process::exit(0);
+    });
```
I guess you can also ensure here that the thread exits gracefully and logs the cause:

```suggestion
    thread::spawn(move || {
        if signals.forever().next().is_some() {
            info!("Shutting down due to signal");
            std::process::exit(0);
        }
    });
```

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,242 @@
+    let mut signals = Signals::new([SIGINT, SIGTERM]).context("Failed to set up signals")?;
+    let inotify = Inotify::init(InitFlags::IN_CLOEXEC).context("Failed to set up inotify")?;
+
+    let source = match args.sentinel_path.parent() {
+        Some(p) if p.is_dir() => p,
+        _ => bail!(
+            "Sentinel ({:?})'s parent is not a directory",
+            args.sentinel_path
+        ),
+    };
```
I guess it can be a bit simplified

```suggestion
    let source = args.sentinel_path.parent().ok_or_else(|| {
        anyhow!("Sentinel ({:?})'s parent is not a directory", args.sentinel_path)
    })?;
```

Saviq replied:
My original code also checks whether `.parent()?.is_dir()`, don't see a clean way to do this without a `match`?

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,242 @@
+            } else {
+                debug!(target: "files", "removed {:?}", &path)
+            }
+        }
```
When iterating over directories, you can use iterators with combinators like filter_map to simplify the code:

```rust
for entry in fs::read_dir(target).context("failed to list target directory")?.filter_map(Result::ok) {
    let path = entry.path();
    if path.is_dir() {
        fs::remove_dir_all(&path).with_context(|| format!("Failed to remove {:?}", path))?;
    } else {
        fs::remove_file(&path).with_context(|| format!("Failed to remove {:?}", path))?;
    }
}
```

Saviq replied:
Looking at this again, it wouldn't allow me to log the debug lines about successful cleanups.

Because of the above (actually making any issues fatal), I did rework this using `with_context()` and simplified quite a bit.

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,242 @@
+}
+
+/// Remove the sentinel file from target, then clear the target directory
+fn cleanup(target: &PathBuf, sentinel_name: &str) -> Result<()> {
```
I'm also a bit concerned that this `cleanup` function performs filesystem operations that might leave the directory in a partially cleaned state. You can consider adding error aggregation or ensuring cleanup's atomicity if possible.

Saviq replied:
Can you elaborate on "error aggregation"?

There isn't really much you can do in this situation, is there… you're trying to recursively remove a directory, so you go through its contents and remove one by one, any of which can fail…

nadzyah replied:
You can use a `Vec<anyhow::Error>` to store all errors that occur during the cleanup. At the end of the function, if any errors are encountered, you return them all wrapped in an `anyhow::Error`

Probably there is a better approach, this is just my idea

Saviq replied:
I don't think it changes much… as is, the errors will be logged one after another, if I aggregate them, same really.

The only real difference I could make here would be to make any cleanup or populate errors fatal. I suppose that would indeed be appropriate.

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,242 @@
+
+        let path = entry.path();
+        let relative_path = path.strip_prefix(source)?;
+        if path == source {
```
This check is redundant since the `WalkDir` iterator already starts at `source`.

Saviq replied:
Well, exactly, and I don't want to do anything to the source? Or do you mean that it starts _below_ the source? Don't think that was what I experienced, but will check.

nadzyah replied:
I guess I got the logic wrong; thanks for the clarification!

nadzyah commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,242 @@
+) -> Result<()> {
+    info!(target: "inotify", "starting event monitoring on {:?}", &source);
+    loop {
+        let events = inotify
+            .read_events()
+            .context("Error reading Inotify events")?;
```
The loop blocks on that `read_events`. If there's no activity, it might be better to implement a timeout or a way to check for shutdown signals.

Saviq replied:
That's the purpose of this service, though… wait for activity in there… I'm not sure if the added complexity adds any value here.

RAOF replied:
I guess the only benefit of having a shutdown signal here would be that, as it stands, stopping this service at the wrong time will leave an operation half-done. The delete-sentinel-first, write-sentinel-last approach will ensure that this *recovers*, but I don't have the context to know if leaving this temporarily in a broken state is acceptable. (I guess if you *really* wanted to avoid a temporary broken state you'd need to do this a bit differently anyway)

Saviq reviewed (commented):
Thanks for the review @nadzyah! Handful of clarifications inline, others I'll get at soon.

Saviq committed:
Review #6: fatal filesystem issues, less PathBuf, more context

Co-authored-by: Nadzeya Hutsko <nadzeya.hutsko@canonical.com>

Saviq reviewed (commented):
@nadzyah thanks for the review, some things I've explained why they are like they are, most applied.

Saviq committed:
kernel-gpu: fix symlink handling

paul-rodriguez reviewed (approved):
After a quick re-read I don't see anything wrong with the Rust code in here.

nadzyah reviewed (approved):
Thanks for addressing my comments and adding an explanation

RAOF commented on component-monitor/src/main.rs:
```diff
@@ -0,0 +1,220 @@
+
+    cleanup(target, sentinel_name).context("Cleanup failed")?;
+    info!(target: "files", "copying from {source:?} to {target:?}");
+    for entry in WalkDir::new(source) {
```
This:
```suggestion
    for entry in WalkDir::new(source).min_depth(1) {
```
would allow you to remove the `if path == source` conditional below (as `source` is the only entry at depth 0)

RAOF reviewed (approved):
This LGTM.

Saviq committed:
Review #7: simplify `WalkDir` use

Co-authored-by: Chris Halse Rogers <christopher.halse.rogers@canonical.com>

Saviq committed:
cargo: update dependencies
